internal/pkg/pluginmanager/pbdownloader.go

```diff
@@ -0,0 +1,119 @@
+// 1. get plugins *.so
+// 2. show progress bar on console
+
+package pluginmanager
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/cheggaaa/pb"
+)
+
+type PbDownloadClient struct {
+	*http.Client
+}
+
+func NewPbDownloadClient() *PbDownloadClient {
+	dClient := PbDownloadClient{}
+	dClient.Client = http.DefaultClient
+	return &dClient
+}
+
+// download from release assets
+func (pd *PbDownloadClient) download(pluginsDir, pluginFilename, version string) error {
+
+	err := createPathIfNotExists(pluginsDir)
+	if err != nil {
+		return err
+	}
+
+	downloadURL := fmt.Sprintf("%s/v%s/%s", defaultReleaseUrl, version, pluginFilename)
+	tmpName := pluginFilename + ".tmp"
+
+	pd.Timeout = time.Second * 60 * 60
+	resp, err := pd.Get(downloadURL)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		log.Printf("[INFO] downloading: [%s]", pluginFilename)
+		log.Print("\n")
+
+		downFile, err := os.Create(filepath.Join(pluginsDir, tmpName))
+		if err != nil {
+			return err
+		}
+
+		defer downFile.Close()
+		// create progress bar when reading response body
+		errSetup := setUpProgressBar(resp, downFile)
+		if errSetup != nil {
+			log.Print(errSetup)
+			return errSetup
+		}
+
+		log.Print("\n")
+		log.Printf("[INFO] [%s] download success.", pluginFilename)
+	} else {
+		log.Print("\n")
+		log.Printf("[ERROR] [%s] download fail,%s.", pluginFilename, resp.Status)
+	}
+
+	// rename, tmp file to real file
+	err = os.Rename(
+		filepath.Join(pluginsDir, tmpName),
+		filepath.Join(pluginsDir, pluginFilename))
+	if err != nil {
+		log.Print(err)
+		return err
+	}
+	return nil
+}
+
+// setUpProgressBar create bar and setup
+func setUpProgressBar(resp *http.Response, downFile *os.File) error {
+	//get size
+	i, _ := strconv.Atoi(resp.Header.Get("Content-Length"))
+	sourceSiz := int64(i)
+	source := resp.Body
+
+	//create a bar and set param
+	bar := pb.New(int(sourceSiz)).SetRefreshRate(time.Millisecond * 10).SetUnits(pb.U_BYTES).SetWidth(100)
+	bar.ShowSpeed = true
+	bar.ShowTimeLeft = true
+	bar.ShowFinalTime = true
+	bar.SetWidth(80)
+	bar.Start()
+
+	writer := io.MultiWriter(downFile, bar)
+	_, err := io.Copy(writer, source)
+	if err != nil {
+		log.Print(err)
+		return err
+	}
+	bar.Finish()
+	return nil
+}
+
+func createPathIfNotExists(path string) error {
+	_, err := os.Stat(path)
+	if err == nil {
```

Comment:
Same here. I'd like to use `if err := func(); err != nil {}`.
There are many places above that have the same problem. Of course, this is not very important, but it is relatively more commonly used.

Reply:
done